test/tlsconnection.js

```diff
@@ -0,0 +1,298 @@
+    // XXX TODO; PROTOCOL_VERSION?
+  });
+
+  it.skip('errors if ClientHello does not offer a compatible cipher suite', async () => {
+    // XXX TODO; HANDSHAKE_FAILURE?
+  });
+
+  it.skip('errors if ClientHello has incorrect supported_versions extension', async () => {
+    // XXX TODO; what error?
+  });
+
+  it.skip('errors if ClientHello does not offer a matching PSK', async () => {
+    // XXX TODO; UNKNOWN_PSK_IDENTITY
+  });
+
+  it.skip('errors if ClientHello does not offer the expected PSK mode', async () => {
+    // XXX TODO; HANDSHAKE_FAILURE
+  });
+
+  it.skip('errors if ClientHello has an invalid PSK binder', async () => {
+    // XXX TODO; DECRYPT_ERROR
+  });
+
+  it.skip('errors if ClientHello has data after the PSK binders', async () => {
+    // XXX TODO; DECODE_ERROR
+  });
+
+  it('accepts a valid ClientHello message with unknown extensions, ignoring them', async () => {
+    // XXX TODO; actually the test-vector ClientHello has unknown extensions already.
+  });
+});
+
+
+describe('the ClientConnection class', () => {
+  let client, CLIENT_SENT;
+
+  beforeEach(async () => {
+    CLIENT_SENT = [];
+    sinon.stub(crypto, 'getRandomValues').callsFake(arr => {
+      arr.set(TEST_VECTORS.CLIENT_RANDOM);
+    });
+    client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+  });
+
+  afterEach(() => {
+    crypto.getRandomValues.restore();
+  });
+
+  it('calls crypto.getRandomValues to generate a strong random salt', async () => {
+    assert.equal(crypto.getRandomValues.callCount, 1);
+  });
+
+  it('immediately sends an initial ClientHello', async () => {
+    assert.equal(CLIENT_SENT.length, 1);
+    // XXX TODO: the generated ClientHello contains extra extensions that
+    // we don't support, can't do a direct comparison here yet...
+    // assert.ok(bytesAreEqual(SERVER_SENT[0], TEST_VECTORS.CLIENT_HELLO));
+  });
+
+  it('errors if it receives non-TLS-handshake-message data', async () => {
+    await assert.throwsAsync(async () => {
+      await client.recv(utf8ToBytes('a badly-formed message'));
+    }, TLSError, 'RECORD_OVERFLOW');
+  });
+
+  it('errors if it receives a message of unknown type', async () => {
+    await assert.throwsAsync(async () => {
+      await client.recv(hexToBytes('0100000002AABB'));
+    }, TLSError, 'UNEXPECTED_MESSAGE');
+  });
+
+  it.skip('errors if it receives a message other than ServerHello', async () => {
+    // XXX TODO
+  });
+
+  it.skip('errors if ServerHello did not select the right CipherSuite', async () => {
+    // XXX TODO
+  });
+
+  it.skip('errors if ServerHello did not select the right PSK', async () => {
+    // XXX TODO
+  });
+
+  it.skip('errors if ServerHello tried to use DHE mode with the PSK', async () => {
+    // XXX TODO
+  });
+
+  it.skip('then accepts a valid ServerHello response', () => {
+    // XXX TODO: feed it the pre-generated ServerHello.
+  });
+});
+
```

Comment:
Some tests to consider, though I'm guessing that you have some of these already:

- test reading and writing more thoroughly, including short reads and long reads
- skipping records fails
- damage the ciphertext and the recipient explodes
- skipping handshake messages fails
- missing extensions means failure
- extra extensions (in ClientHello are ignored, in ServerHello generate alerts)
- test that writing blocks/fails prior to the handshake being complete (you could write at the server in 0.5-RTT, but please disable that and test that it is disabled)
- test that you ignore ChangeCipherSpec (or that you choke cleanly) and limit it to during the handshake if you do accept it, if you do ignore it, the server should probably emit one if the client sends a non-zero session ID
- test that you are actually encrypting (yep, that's somewhat tricky, but it's pretty important)
- test without a mocked RNG and see that you get different ClientHello/ServerHello when you run the handshake multiple times
- check that reordered extensions still work fine
- add dummy versions, ciphersuites, and PSK modes to extensions and see that these are ignored
- add a non-zero compression mode and the server should explode
- validate that the version in the record header is correct (TLS 1.0 I think is right: 0x0301)
- validate the version in the ServerHello is TLS 1.2 (0x0303)
- handle padded records properly
- abort if a record only contains zeros (i.e., it is over-padded)
- abort if a handshake record is empty (i.e., it is properly padded with a handshake type, but no data)
- abort if an alert of any kind is received

Reply:
> you could write at the server in 0.5-RTT

To clarify, do you mean the server sending application data after it has sent its `Finished`, but before it has received `Finished` from the client in return?

Reply:
Right.  In TLS 1.3, the server can send before it receives the Finished from the client, but that is generally considered risky in mutually authenticated modes.  For your purposes, you want to wait until you have verified the peer's Finished, in both directions.